samples/crypto/tls/src/tls_functions_client.c

```diff
@@ -0,0 +1,181 @@
+{
+	ssize_t sent_len;
+
+	while (len > 0) {
+		sent_len = send(sock, buf, len, 0);
+		if (sent_len < 0) {
+			return sent_len;
+		}
+
+		buf = (const uint8_t *)buf + sent_len;
+		len -= sent_len;
+	}
+
+	return 0;
+}
+
+
+/** @brief Function for setting up the client socket to use for TLS handshake.
+ *
+ * @returns The socket identifier.
+ */
+static int setup_tls_client_socket(void)
+{
+	int err;
+	int sock;
+	int verify = REQUIRED;
+
+	/* List of security tags to register. */
+	sec_tag_t sec_tag_list[] = {
+		CA_CERTIFICATE_TAG,
+		#if defined(CONFIG_MBEDTLS_KEY_EXCHANGE_PSK_ENABLED)
+		PSK_TAG,
+		#endif
+	};
+
+	sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TLS_1_2);
+	if (sock < 0) {
+		LOG_ERR("Failed to aquire socket. Err: %d", errno);
+		return -errno;
+	}
+
+	err = setsockopt(sock, SOL_TLS, TLS_PEER_VERIFY,
+			 &verify, sizeof(verify));
+	if (err < 0) {
+		LOG_ERR("Failed to set TLS peer verification. Err %d", errno);
+		return -errno;
+	}
+
+	err = setsockopt(sock, SOL_TLS, TLS_SEC_TAG_LIST, sec_tag_list,
+			 sizeof(sec_tag_list));
+	if (err < 0) {
+		LOG_ERR("Failed to set TLS security TAG list. Err: %d", errno);
+		return -errno;
+	}
+
+	err = setsockopt(sock, SOL_TLS, TLS_HOSTNAME,
+			 TLS_PEER_HOSTNAME, sizeof(TLS_PEER_HOSTNAME));
+	if (err < 0) {
+		LOG_ERR("Failed to set TLS_HOSTNAME option. Err: %d", errno);
+		return -errno;
+	}
+
+	return sock;
+}
+
+/** @brief Thread function that handles all the steps of the TLS handshake for
+ *  a client. Loops forever and connects on given port number.
+ *
+ */
+static void process_tls_client(void)
+{
+	int ret;
+	int sock;
+	int received;
+	struct sockaddr_in server_addr;
+	socklen_t server_addr_len = sizeof(server_addr);
+
+	server_addr.sin_family = AF_INET;
+	server_addr.sin_port = htons(SERVER_PORT);
+	inet_pton(AF_INET, CONFIG_NET_CONFIG_PEER_IPV4_ADDR,
+		  &server_addr.sin_addr);
+
+	while (true) {
+
+		sock = setup_tls_client_socket();
+		if (sock < 0) {
+			LOG_INF("Retrying to aquiring socket");
+			k_sleep(K_MSEC(1000));
+			continue;
+		}
+
+		LOG_INF("Initiates TLS connection on port %d ...",
+			SERVER_PORT);
+		ret = connect(sock, (struct sockaddr *) &server_addr,
+			      server_addr_len);
+
+		if (ret < 0) {
+			LOG_WRN("Cannot make TLS connection. Err: %d", errno);
+			LOG_INF("Will re-try connection in 1 second");
+			(void)close(sock);
+			k_sleep(K_MSEC(1000));
+			continue;
+		}
+		LOG_INF("TLS handshake completed");
+
+		while (true) {
+			// Try to receive data while connected
+			received = recv(sock, m_recv_buffer,
+					sizeof(m_recv_buffer), 0);
+
+			if (received == 0) {
+				/* Connection closed */
+				LOG_INF("TLS Connection closed");
+				break;
+
+			} else if (received < 0) {
+				/* Socket error */
+				LOG_ERR("TLS Connection error: %d", errno);
+				break;
+			}
+
+			LOG_INF("TLS received %d byte(s)", received);
+
+			ret = send_buffer(sock, m_recv_buffer, received);
+			if (ret != 0) {
+				LOG_ERR("TLS Failed to send");
```

Comment:
Would be good to log the actual error code returned.

Reply:
fixed